src/frontend/src/optimizer/rule/pull_up_correlated_predicate.rs

```diff
@@ -0,0 +1,136 @@
+use crate::optimizer::PlanRef;
+use crate::utils::Condition;
+
+/// This rule is for pattern: Apply->Project->Filter.
+///
+/// To unnest, we just pull predicates contain correlated variables in Filter into Apply, and
+/// convert it into corresponding type of Join.
+pub struct PullUpCorrelatedPredicate {}
+impl Rule for PullUpCorrelatedPredicate {
+    fn apply(&self, plan: PlanRef) -> Option<PlanRef> {
+        let apply = plan.as_logical_apply()?;
+
+        let right = apply.right();
+        let project = right.as_logical_project()?;
+        let (mut exprs, mut expr_alias, _) = project.clone().decompose();
+        let begin = exprs.len();
+
+        let input = project.input();
+        let filter = input.as_logical_filter()?;
+
+        // Split predicates in LogicalFilter into correlated expressions and uncorrelated
+        // expressions.
+        let (cor_exprs, uncor_exprs) =
+            filter
+                .predicate()
+                .clone()
+                .into_iter()
+                .partition_map(|expr| {
+                    let mut rewriter = Rewriter {
+                        has_correlated_input_ref: false,
+                        input_refs: vec![],
+                        index: exprs.len() + apply.left().schema().fields().len(),
+                    };
+                    let rewritten_expr = rewriter.rewrite_expr(expr.clone());
+                    if rewriter.has_correlated_input_ref {
+                        // Append input_refs in expression which contains correlated_input_ref to
+                        // `exprs` used to construct new LogicalProject.
+                        exprs.extend(
+                            rewriter
+                                .input_refs
+                                .drain(..)
+                                .map(|input_ref| input_ref.into()),
+                        );
+                        Either::Left(rewritten_expr)
+                    } else {
+                        Either::Right(expr)
+                    }
+                });
+
+        // TODO: remove LogicalFilter with always true condition.
+        let filter = LogicalFilter::new(
+            filter.input(),
+            Condition {
+                conjunctions: uncor_exprs,
+            },
+        );
+
+        // Add columns involved in expressions removed from LogicalFilter to LogicalProject.
+        let end = exprs.len();
+        expr_alias.extend(vec![None; end - begin].into_iter());
+
+        let project = LogicalProject::new(filter.into(), exprs, expr_alias);
+
+        // Merge these expressions with LogicalApply into LogicalJoin.
+        let on = Condition {
+            conjunctions: cor_exprs,
+        };
+        Some(LogicalJoin::new(apply.left(), project.into(), apply.join_type(), on).into())
+    }
+}
+
+/// Rewritter is responsible for rewriting `correlated_input_ref` to `input_ref` and shifting
+/// `input_ref`.
+struct Rewriter {
```

Comment:
```suggestion
/// Rewrites a pulled predicate expression. It is pulled from the right of the apply to the `on` clause.
///
/// Rewrites `correlated_input_ref` (referencing left side) to `input_ref` and shifting `input_ref` (referencing right side).
///
/// Also collects all `InputRef`s, which will be added to the project, so that they are accessible by the expression after it is pulled.
struct Rewriter {
```